src/rust/cryptography-x509/src/common.rs

```diff
@@ -301,9 +301,72 @@ impl<'a> asn1::SimpleAsn1Writable for UnvalidatedVisibleString<'a> {
     }
 }
 
+/// Like `UnvalidatedIA5String`, but preserves the invariant that the
+/// underlying string is ASCII only.
+#[derive(Debug, PartialEq)]
+pub struct IA5String(String);
+
+impl IA5String {
+    pub(crate) fn new(value: String) -> Option<Self> {
+        if value.is_ascii() {
+            Some(Self(value))
+        } else {
+            None
+        }
+    }
+
+    pub(crate) fn as_str(&self) -> &str {
+        &self.0
+    }
+}
+
+/// A `DNSString` is an `IA5String` with additional invariant preservations
+/// per RFC 5280 4.2.1.6.
+///
+/// In particular, a `DNSString` is normalized to lowercase ASCII internally
+/// and cannot contain `" "`.
+#[derive(Debug, PartialEq)]
+pub struct DNSString(IA5String);
+
+impl DNSString {
+    pub fn new(value: &str) -> Option<Self> {
+        if value.is_empty() || value == " " {
+            None
+        } else {
+            IA5String::new(value.to_lowercase()).map(Self)
+        }
+    }
+
+    pub fn as_str(&self) -> &str {
+        self.0.as_str()
+    }
+
+    pub fn matches(&self, pattern: &Self) -> bool {
+        let hostname = self.as_str();
+        let pattern = pattern.as_str();
+
+        match (hostname.split_once('.'), pattern.split_once('.')) {
+            // If both hostname and pattern contain multiple labels, then
+            // we attempt to match using a subset of RFC 6125 6.4.3.
+            // In particular, we don't attempt to support anything
+            // except left-most wildcards.
+            (Some((subdomain, parent)), Some((pat_subdomain, pat_parent))) => {
+                // TODO: This is almost certainly insufficient: we also need
+                // to check for nonsense patterns like `*.`.
+                (pat_subdomain == "*" || pat_subdomain == subdomain) && pat_parent == parent
```

Comment:
IMO it makes sense to round this out before merging, but I opened this PR with the TODO still in place in case @alex or @reaperhulk would prefer to merge an MVP first.